assets/js/util/tracking/createTrackEvent.js

```diff
@@ -32,17 +33,49 @@ export default function createTrackEvent( config, dataLayerTarget ) {
 		} = config;
 
 		if ( ! trackingEnabled ) {
-			return;
+			// Resolve immediately if tracking is disabled.
+			return Promise.resolve();
 		}
 
-		dataLayerPush( 'event', eventName, {
-			send_to: trackingID,
-			event_category: eventCategory,
-			event_label: eventLabel,
-			event_value: eventValue,
-			dimension1: referenceSiteURL,
-			dimension2: isFirstAdmin ? 'true' : 'false',
-			dimension3: userIDHash,
+		return new Promise( ( resolve ) => {
+			const eventData = {
+				send_to: trackingID,
+				event_category: eventCategory,
+				event_label: eventLabel,
+				event_value: eventValue,
+				dimension1: referenceSiteURL,
+				dimension2: isFirstAdmin ? 'true' : 'false',
+				dimension3: userIDHash,
+			};
+
+			let resolved = false;
+			const resolveOnce = ( eventSent ) => {
+				if ( resolved ) {
+					return;
+				}
+				resolved = true;
+				if ( ! eventSent ) {
+					global.console.warn( `Tracking event "${ eventName }" (category "${ eventCategory }") took too long to fire.` );
+				}
+				resolve();
+			};
+
+			// This timeout ensures a tracking event does not block the user
+			// event if it is not sent (in time).
+			// If this fails, it shouldn't reject the promise since event
+			// tracking should not result in user-facing errors. It will just
+			// trigger a console warning.
+			const failTimeout = setTimeout( () => {
```

Comment:
I intentionally didn't put it there because `resolveOnce` ensures that it's only called once; if the timeout is hit where before the `event_callback` was called, `resolveOnce` should still only be run once due to its cancel logic in the beginning. We don't wanna trigger the console warning if the `event_callback` was actually called before the timeout.